includes/class-integrate-convertkit-wpforms-setup.php

```diff
@@ -0,0 +1,197 @@
+<?php
+/**
+ * Plugin activation, update and deactivation class.
+ *
+ * @package Integrate_ConvertKit_WPForms
+ * @author  ConvertKit
+ */
+
+/**
+ * Runs any steps required on plugin activation, update and deactivation.
+ *
+ * @package Integrate_ConvertKit_WPForms
+ * @author  ConvertKit
+ * @version 1.5.0
+ */
+class Integrate_ConvertKit_WPForms_Setup {
+
+	/**
+	 * Runs routines when the Plugin version has been updated.
+	 *
+	 * @since   1.5.0
+	 */
+	public function update() {
+
+		// Get installed Plugin version.
+		$current_version = get_option( 'integrate_convertkit_wpforms_version' );
+
+		// If the version number matches the plugin version, no update routines
+		// need to run.
+		if ( $current_version === INTEGRATE_CONVERTKIT_WPFORMS_VERSION ) {
+			return;
+		}
+
+		/**
+		 * 1.5.0: Migrate API settings from individual forms to integrations.
+		 */
+		if ( ! $current_version || version_compare( $current_version, '1.5.0', '<' ) ) {
+			$this->migrate_settings_from_forms_to_integrations();
+		}
+
+		// Update the installed version number in the options table.
+		update_option( 'integrate_convertkit_wpforms_version', INTEGRATE_CONVERTKIT_WPFORMS_VERSION );
+
+	}
+
+	/**
+	 * 1.5.0+: For each WPForms Form that has ConvertKit settings defined (API Key, Form ID etc), migrate
+	 * those settings to a provider connection (WPForms > ), and configures the WPForms Form to use the
+	 * integration connection instead of the form settings.
+	 *
+	 * @since   1.5.0
+	 */
+	private function migrate_settings_from_forms_to_integrations() {
+
+		// Bail if the WPForms handler class isn't available.
+		if ( ! class_exists( 'WPForms_Form_Handler' ) ) {
+			return;
+		}
+
+		// Get all forms.
+		$form_handler = new WPForms_Form_Handler();
+		$forms        = $form_handler->get();
+
+		// Bail if no WPForms Forms exist.
+		if ( ! is_array( $forms ) ) {
+			return;
+		}
+		if ( count( $forms ) === 0 ) {
+			return;
+		}
+
+		// For each form, inspect its settings to determine if < 1.5.0 ConvertKit settings exist in the Form.
+		foreach ( $forms as $form ) {
+			// Decode settings into array.
+			$data = wpforms_decode( $form->post_content );
+
+			// Skip if this Form already has a ConvertKit provider configured.
+			// This shouldn't be the case for upgrades to 1.5.0+, but it's a useful
+			// sanity check.
+			if ( array_key_exists( 'providers', $data ) && array_key_exists( 'convertkit', $data['providers'] ) ) {
+				continue;
+			}
+
+			// Bail if the API Key or Form ID settings don't exist or are empty.
+			if ( ! array_key_exists( 'be_convertkit_api', $data['settings'] ) ) {
```

Comment:
Can this plugin take multiple CK accounts (api keys)? Regardless, it might make sense to have caching since presumably the account doesn't need to be found/created for every form, and depending on the cost of the lookup, it could bog this migration down.

Reply:
@noelherrick Yes, it can take multiple ConvertKit accounts.  This migration runs once, to move ConvertKit API Keys that might be stored in each form to the WPForms provider/integration model.  There's a check here that will either:
- register a new ConvertKit connection in the WPForms provider/integration model, if one does not already exist that uses this given API Key, or
- use the existing ConvertKit connection in the WPForms provider/integration model, if one already exists using this given API Key.

It's an inexpensive lookup, and at no point in performing this (and/or creating/assigning a connection) do we hit the ConvertKit API itself.  It's a one off migration to move where we store ConvertKit settings, basically.

[This PR](https://github.com/ConvertKit/convertkit-wpforms/pull/4) has a handy guide covering the before and after of where the ConvertKit settings are stored, which probably gives better context and explains things a bit more clearly.